source/extensions/filters/http/ext_authz/ext_authz.cc

```diff
@@ -208,6 +208,15 @@ void Filter::onComplete(Filters::Common::ExtAuthz::ResponsePtr&& response) {
   using Filters::Common::ExtAuthz::CheckStatus;
   Stats::StatName empty_stat_name;
 
+  if (response->status != CheckStatus::Error) {
```

Comment:
Is there anything about an Error status that would make the `response->dynamic_metadata` not valid?

I would think that if the dynamic_metadata set is non-empty that means it was explicitly supplied, so it wouldn't matter whether the check status error'd or not. Wouldn't this cover the use case of a gRPC service returning some sort of error but still wanting to propagate metadata, perhaps for logging purposes? What do you think?

Reply:
I ask because we recently added dynamic_metadata support to `ratelimit` here https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/ratelimit/ratelimit.cc#L155. This sets the dynamic metadata regardless of status. It might also be interesting to hear from @gbrail on how ext_proc wants to implement dynamic metadata since I'm thinking that ext_authz/ratelimit/proc will all converge into something very similar before long.

Reply:
I think `CheckStatus::Error` for ext_authz implementation means failing on getting a response from the authorization server. I could be wrong: https://github.com/envoyproxy/envoy/blob/7adc0393221467fcae18052c1f7caf0cb6048d2e/source/extensions/filters/common/ext_authz/ext_authz_grpc_impl.cc#L96-L97 but looking at that line seems like we have no access to the corresponding envoy::service::auth::v3::CheckResponse from the server.

Reply:
It's definitely true that we create a synthetic response with an Error status on it when we fail to get a response from the authorization server. Looking harder, it seems like we already map grpcStatus != OK -> CheckStatus::Denied, so my original concern about "failed response, but with metadata" is not valid right now.

I still think for consistency it makes sense to treat the dynamic metadata set as the source of truth for whether or not it should be set, but knowing that it's not actually an issue today makes me less worried about it.

Reply:
Don't we require a response from the ext authz server to get the metadata to set, and error means that we didn't get one? Or is the metadata coming from somewhere else?

Reply:
We do require a response for dynamic metadata, and an error means we didn't get one. But my point was mostly that this is an implementation detail of the caller and IMO it would be more intuitive if the dynamic metadata object itself was the source of truth for whether there exists dynamic metadata to set. This way it becomes theoretically possible to create callers that pass an Error status but still want to propagate some metadata.

I don't think this is a blocker but it would be nice to have consistency with the way the ratelimit (and ext_proc) handles this scenario.

Reply:
I'm OK either way. I can take out the test for skipping emitting data when `CheckStatus::Error`. However, my slight preference is to keep it like this since currently, having non-`CheckStatus::Error`-response is the only way to get the metadata being filled. We can always change this behavior if that's no longer true.

Reply:
I think it's probably fine to not check for Error here, since we should only get a non-empty metadata field if we actually got a response from the server. There seems to be no way to make the ext authz client propagate an Error from the authz server (lots of conversion logic in ext_authz_grpc_impl.cc), so I'm not sure setting metadata on Error actually permits more flexibility.

Reply:
Gotcha.

Reply:
Updated.